app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,208 @@
+
+        filter_term = 'Chips'
+        response = self.client.get(
+            RECIPES_URL,
+            {'name': filter_term}
+        )
+
+        # Then
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data), 0)
+
+    def test_get_recipe_by_id_success(self):
+        """Test retrieving a recipe by unique ID"""
+
+        sample_recipe()
+        recipe = sample_recipe({
+            'name': 'Bangers and Mash',
+            'description': 'A hearty meal'
+        })
+        recipe.ingredients.add(sample_ingredient(recipe=recipe, name='Mash'))
+
+        serializer = RecipeSerializer(recipe)
+
+        url = recipe_url(recipe.id)
+        response = self.client.get(url)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data, serializer.data)
+
+    def test_get_recipe_by_id_not_found(self):
+        """Test retrieving a recipe by ID when the recipe doesn't exist"""
+
+        response = self.client.get(recipe_url(recipe_id=1))
+
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_create_recipe(self):
+        """Test creating a new recipe"""
+
+        ingredient_name_1 = 'Sausage'
+        ingredient_name_2 = 'Gravy'
+
+        payload = {
+            'name': 'Bangers And Mash',
+            'description': 'Whack it all together',
+            'ingredients': [
+                {'name': ingredient_name_1},
+                {'name': ingredient_name_2},
+            ]
+        }
+
+        response = self.client.post(RECIPES_URL, payload, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        recipe = Recipe.objects.get(id=response.data['id'])
+        ingredients = recipe.ingredients.all()
+
+        self.assertEqual(ingredients.count(), 2)
+
+    def test_create_recipe_invalid_request(self):
+        """Test creating a new recipe with invalid payload"""
+
+        payload = {
+            'name': '',
+        }
+
+        response = self.client.post(RECIPES_URL, payload)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_patch_recipe(self):
+        """Test updating of a recipe"""
+
+        recipe = sample_recipe()
+
+        ingredient1 = sample_ingredient(recipe, name='Garlic')
+        ingredient2 = sample_ingredient(recipe, name='Cheese')
+
+        recipe.ingredients.add(ingredient1, ingredient2)
+
+        new_ingredient_name = 'Egg'
+        payload = {
+            'name': recipe.name,
+            'description': recipe.description,
+            'ingredients': [
+                {'name': new_ingredient_name}
+            ]
+        }
+
+        response = self.client.patch(
+            recipe_url(recipe_id=recipe.id), payload, format='json'
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        recipe = Recipe.objects.get(id=response.data['id'])
```

Comment:
Similar to the query above, you could do this in one query.

Furthermore, doing
```
ingredient_names = ingredients.values_list("name", flat=True)
```
will do `SELECT name FROM ..` instead of `SELECT * FROM`

Premature improvement at this point, but so you know :wink: